Suggestion for new dynamic detachable joint plugin

AdarshKaran opened the pull request:
<!--
Please remove the appropriate section.
For example, if this is a new feature, remove all sections except for the "New feature" section

If this is your first time opening a PR, be sure to check the contribution guide:
https://gazebosim.org/docs/all/contributing
-->

# 🎉 New feature

Closes #2362

Depends on https://github.com/gazebosim/gz-msgs/pull/533

## Summary
Created a `dynamic_detachable_joint` system as an extension of the detachable joint system where the `childModelName` and `childLinkName` can now be passed through a gz service or a ROS2 service (a bridge has also been added for this). A `command`: `attach` or `detach` can be specified in the service call. I have also added an `attachDistance` option that allows you to specify how far the child should be to be considered attached.

Most of the code is based on the already existing DetachableJoint Plugin logic

## Test it

Adding the plugin to robot model URDF/SDF
```
    <gazebo>
      <plugin filename="gz-sim-dynamic-detachable-joint-system"
            name="gz::sim::systems::DynamicDetachableJoint">
        <parent_link>robotiq_85_left_finger_tip_link</parent_link>
        <service_name>/payload/attach_detach</service_name>
        <output_topic>/child_state</output_topic>
        <attach_distance>0.25</attach_distance>
      </plugin>
    </gazebo>

```

GZ Service call command
```
# Attach
gz service -s /payload/attach_detach \
    --reqtype gz.msgs.AttachDetachRequest \
    --reptype gz.msgs.AttachDetachResponse \
    --timeout 3000 \
    --req 'child_model_name: "cube", 
           child_link_name: "link", 
           command: "attach"'

# Detach
gz service -s /payload/attach_detach \
    --reqtype gz.msgs.AttachDetachRequest \
    --reptype gz.msgs.AttachDetachResponse \
    --timeout 2000 \
    --req 'child_model_name: "cube", 
           child_link_name: "link", 
           command: "detach"'
```
ROS2 Service call command
```
# Attach
ros2 service call /payload/attach_detach \
    ros_gz_interfaces/srv/AttachDetach \
    "{child_model_name: 'cube', 
      child_link_name: 'link', 
      command: 'attach'}"

# Detach  
ros2 service call /payload/attach_detach \
    ros_gz_interfaces/srv/AttachDetach \
    "{child_model_name: 'cube', 
      child_link_name: 'link', 
      command: 'detach'}"
```

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers
- [x] Was GenAI used to generate this PR? If so, make sure to add "Generated-by" to your commits. (See [this policy](https://osralliance.org/wp-content/uploads/2025/05/OSRF-Policy-on-the-Use-of-Generative-Tools-Generative-AI-in-Contributions.pdf) for more info.) Yes. Chat GPT in copilot  was used partly for help with formatting and for the first draft of some of the comments. 

Generated-by: Remove this if GenAI was not used.

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` and `Generated-by` messages.

GIF

![Dynamic detachable joint plugin demonstration Full](https://github.com/user-attachments/assets/908fe96b-5110-4a37-a87e-a0ff78ab1421)
*The cylindrical cans are being attached to the gripper using a gripper bridge relay node which does the service call*

ahcorde reviewed (changes requested):
It will be great to add some details in the header file as documentation

arjo129 commented:
Thanks for this contribution. @AdarshKaran please sign off your commits. You can see how here:  https://github.com/gazebosim/gz-sim/pull/3086/checks?check_run_id=50152759632

AdarshKaran commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,383 @@
+using namespace systems;
+
+/////////////////////////////////////////////////
+void DynamicDetachableJoint::Configure(const Entity &_entity,
```
https://github.com/gazebosim/gz-sim/blob/361af45888b96fd860f56c9c2c785fe61e339113/src/systems/dynamic_detachable_joint/DynamicDetachableJoint.hh#L40C1-L64C64

I have added these comments in the header file similar to what was done in the existing DetachableJoint.hh plugin. If you could clarify if i need to add more detailed comments with usage example in the .cc source file above the Configure function then I shall do that.

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+              << std::endl;
+      }
+      else
+      {
```
Knit: I recommend applying the error-first pattern (also called "happy path" programming) here. By checking for and returning early on error conditions, we can significantly reduce code nesting and enhance the overall readability.

Reference: https://preslav.me/2023/09/22/ditch-that-else/

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+               << " could not be found.\n";
+      }
+    }
+    else if (!this->suppressChildWarning)
```
Knit: Why do we need this. Can't we just rely on gazebo's verbosity level?

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+     if (this->isAttached) 
+     {
+       _res.set_success(false);
+      _res.set_message("Already attached to child model " + this->attachedChildModelName +
```
Knit: Indentation issue

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+  gz::msgs::StringMsg detachedStateMsg;
+  if (attached)
+  {
+    detachedStateMsg.set_data("attached to " +
```
Why are we publishing a custom message with custom semantics? I'm not sure we should publish this at all.

arjo129 replied:
`DetachableJoint.cc` publishes only 2 types of outputs: "attached" or "detached", in your case you publish "attached to ..." (if I were redisigning this I'd just publish a bool). If I were writing a parser to interact with your software I'd have to look up this magic string. Its much better to either publish an empty string or just the linked name so end users dont have to write custom parsers. Alternatively publish the entity/knullentity that the object is attached to.

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+ }
+
+//////////////////////////////////////////////////
+void DynamicDetachableJoint::OnAttachRequest(const msgs::Empty &)
```
Why are we passing an empty message for no reason? We don't actually need this method.

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+    this->childModelName = _req.child_model_name();
+    this->childLinkName  = _req.child_link_name();
+     this->OnAttachRequest(msgs::Empty());
+     _res.set_success(true);
```
knit: Indentation

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+    // set the child model and link names from the request
+    this->childModelName = _req.child_model_name();
+    this->childLinkName  = _req.child_link_name();
+     this->OnAttachRequest(msgs::Empty());
```
directly assign the `this->attachRequested = true;` without calling the method,

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+     }
+
+    // set the child model and link names from the request
+    this->childModelName = _req.child_model_name();
```
Knit: I don't think atomics are sufficient here. You do need to use a lock to guarantee data consistency between
`this->childModelName`, `this->childModelName` and `this->attachRequested`.

arjo129 commented on src/systems/dynamic_detachable_joint/DynamicDetachableJoint.cc:
```diff
@@ -0,0 +1,386 @@
+     }
+
+    // Validate that the request matches what is actually attached.
+    const auto &reqModel = _req.child_model_name();
```
Do we actually need to perform this validation?

arjo129 reviewed (changes requested):
Thanks for the contribution! This can be super useful.

I'm being a bit picky, but: 
(1) We should at least have an integration test for this system.
(2) Check indentation + Address the comments I've left for you.
(3) You may need to use a lock instead of atomics as we are configuring multiple parameters at once.

AdarshKaran commented:
There is actually one issue. When i give attach request and specify a `child_model_name` that doesn't exist in the scene, we get a true response from the service. The main reason is because we can't access EntityComponentManager within the OnServiceRequest function. Do you think there is a better way to do it? or shall we just update the response message to something like "Attach request accepted" instead of the current "Attached to child model..."?

arjo129 commented:
> There is actually one issue. When i give attach request and specify a `child_model_name` that doesn't exist in the scene, we get a true response from the service. The main reason is because we can't access EntityComponentManager within the OnServiceRequest function. Do you think there is a better way to do it? or shall we just update the response message to something like "Attach request accepted" instead of the current "Attached to child model..."?

Yes, this is one of the less-than-ideal side effects of the way we mix pub/sub with ECM in Gazebo. I think "Attach request accepted" may be a reasonable thing and then log an error. If need be we could publish to a seperate topic, but I'd leave that as "future work".

AdarshKaran commented:
> > There is actually one issue. When i give attach request and specify a `child_model_name` that doesn't exist in the scene, we get a true response from the service. The main reason is because we can't access EntityComponentManager within the OnServiceRequest function. Do you think there is a better way to do it? or shall we just update the response message to something like "Attach request accepted" instead of the current "Attached to child model..."?
> 
> Yes, this is one of the less-than-ideal side effects of the way we mix pub/sub with ECM in Gazebo. I think "Attach request accepted" may be a reasonable thing and then log an error. If need be we could publish to a seperate topic, but I'd leave that as "future work".

Done! Please let me know if it looks good. I have also added some additional logs and errors for more clarity

arjo129 reviewed (commented):
I have a quick question here. It seems like I forgot to update the review.

In general the code _looks_ ok but I cannot approve it without examples/tests.

arjo129 commented:
Hey @AdarshKaran I havent had time to look into  this yet. I'll get back to you by the end of the week.

AdarshKaran commented:
> Hey @AdarshKaran I havent had time to look into this yet. I'll get back to you by the end of the week.

Hi @arjo129 any update on this?

AdarshKaran commented:
> We discussed this PR last week at the Gazebo Asia sync. From my end it's good to go. @azeey wants to go through the end to end integration once - particularly as we are adding a message for this.

That's great! Makes sense @arjo129 !

arjo129 commented:
Go for it.

AdarshKaran commented:
Hi @azeey @arjo129 @ahcorde, I’ve pushed updates after the gz-msgs changes:

  - DynamicDetachableJoint uses gz.msgs.Result as the service response.
  - Request command also uses the AttachDetachRequest enum instead of a string.
  - Updated the integration test, test world, and example world service calls.
  - Added a DynamicDetachableJoint tutorial and cross reference from the DetachableJoint tutorial
  - Added a GIF to the PR description.

arjo129 commented:
This might be a bit of an ask: can we retarget main? You may need to cherry pick your commits/rebase so that the history is not wrecked.

AdarshKaran committed:
Added dynamic detachable joint system

Signed-off-by: Adarsh Karan Kesavadas Prasanth <adarshkaran01@gmail.com>

arjo129 commented on tutorials/detachable_joints.md:
```diff
@@ -54,7 +58,7 @@ used as the parent link in the detachable joint.
 joint.
 
-* `<child_model_link>`:  Name of the link in the `<child_model>` that will be used
+* `<child_link>`:  Name of the link in the `<child_model>` that will be used
```
Nice catch! Thanks.

arjo129 commented on tutorials/dynamic_detachable_joints.md:
````diff
@@ -0,0 +1,82 @@
+`ros_gz_interfaces/srv/AttachDetach`:
+
+```bash
+ros2 service call /payload/attach_detach \
````
I havent yet reviewed the rosbridge PR.

arjo129 reviewed (approved):
Tested this and it works for me! Thank you. I still need to review the rosbridge PR.  Haven't had the time to do that. There is also a minor comment upstream that may need addressing in the message format.

AdarshKaran commented:
Hey @arjo129! any updates on when this might be merged? or is there anything else pending from my side?

arjo129 commented:
Thanks for the reminder this has fallen of my radar. I think the message
files need to be merged then I'll merge this once the CI goes green.

On Fri, Jun 5, 2026 at 12:42 PM Adarsh Karan Kesavadas Prasanth <
***@***.***> wrote:

> *AdarshKaran* left a comment (gazebosim/gz-sim#3086)
> <https://github.com/gazebosim/gz-sim/pull/3086#issuecomment-4628265393>
>
> Hey @arjo129 <https://github.com/arjo129>! any updates on when this might
> be merged? or is there anything else pending from my side?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/gazebosim/gz-sim/pull/3086?email_source=notifications&email_token=AAEEMQCHRNOJG3SXY6LLV5346JFUFA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTINRSHAZDMNJTHEZ2M4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-4628265393>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEEMQGEHLBYVMKAO4UV3SL46JFUFAVCNFSM6AAAAACGH77LR2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DMMRYGI3DKMZZGM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

arjo129 committed:
Merge branch 'main' into plugin/dynamic_detachable_joint

AdarshKaran commented:
> Thanks for the reminder this has fallen of my radar. I think the message files need to be merged then I'll merge this once the CI goes green.
> […](#)
> On Fri, Jun 5, 2026 at 12:42 PM Adarsh Karan Kesavadas Prasanth < ***@***.***> wrote: *AdarshKaran* left a comment ([gazebosim/gz-sim#3086](https://github.com/gazebosim/gz-sim/pull/3086)) <[#3086 (comment)](https://github.com/gazebosim/gz-sim/pull/3086#issuecomment-4628265393)> Hey @arjo129 <https://github.com/arjo129>! any updates on when this might be merged? or is there anything else pending from my side? — Reply to this email directly, view it on GitHub <#3086?email_source=notifications&email_token=AAEEMQCHRNOJG3SXY6LLV5346JFUFA5CNFSNUABFM5UWIORPF5TWS5BNNB2WEL2JONZXKZKDN5WW2ZLOOQXTINRSHAZDMNJTHEZ2M4TFMFZW63VHNVSW45DJN5XKKZLWMVXHJLDGN5XXIZLSL5RWY2LDNM#issuecomment-4628265393>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAEEMQGEHLBYVMKAO4UV3SL46JFUFAVCNFSM6AAAAACGH77LR2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DMMRYGI3DKMZZGM> . You are receiving this because you were mentioned.Message ID: ***@***.***>

hey! any updates on when this will be merged?

arjo129 committed:
Merge branch 'main' into plugin/dynamic_detachable_joint

arjo129 commented:
https://github.com/gazebosim/gz-sim/actions/runs/30355356768/job/90262262850#step:5:1

Style checks are still not passing.
